File: graphs/CTCI_4.3_list_of_depths.js
// // 4.3 Given a binary tree, design an algorithm which creates a linked list
// of all the nodes at each depth (eg, if you have a tree with depth D, you'll have D linked lists)

// misc:
// each node has a link to its parent.
// any kind of tree
// add children to a queue

function linkNodes(node) {
  level = 1;
  linkedNodes = {};
  q = new Queue();
  q.enq({node, level})
  while (!q.isEmpty()) {
    level = q.peek().level
    if(!linkedNodes[level]) {
      // create new list for new level
      linkedNodes[level] = new LinkedList();
    }
    // enq children
    currNode = q.deq
    if(currNode.left) {
      q.enq(currNode.left, level + 1)
    }
    if(currNode.right) {
      q.enq(currNode.right, level + 1)
    }
    linkedNodes[level].add(currNode);
  }
  return linkedNodes;
}

File: graphs/CTCI_4.3_list_of_depths.rb
## 4.3 Given a binary tree, design an algorithm which creates a linked list
# of all the nodes at each depth (eg, if you have a tree with depth D, you'll have D linked lists)

# misc:
# time complexity:
# space complexity
# # worst: imbalanced tree
# # avg:

def link_nodes(n)
  list_array = []
  parent = Queue.new
  child = Queue.new

  parent.enq(n)
  list_a = LinkedList.new
  while !parent.empty?
    # same as pop
    temp = parent.deq
    list_a.append_after(temp)
    child.enq(temp.left)
    child.enq(temp.right)
    if !parent.empty?
      tempQ = parent
      parent = child
      child = tempq
    else
      list_array << list_a
      list_a = LinkedList.new
    end
  end
  return list_array
end

# recursive solution
def level_lists(tree)
  height = tree.height
  lists = []
  # i current level
  height.tree do |i|
    current_level = LinkedList.new
    helper(current_level, i, tree.root)
    lists << current_level
  end
end
# TODO more descriptive name for helper
def recursive_helper(list, level_num, node)
  if num == 0
    list.add(node)
  end
  # TODO check if left and right nodes exist
  helper(list, num-1, node.left)
  helper(list, num-1, node.right)
end
